src/app.ts

```diff
@@ -487,74 +496,100 @@ export class App {
+
         forEachOrdered(resource.methods, method => {
           if (method.description) {
             out.comment(formatComment(method.description));
           }
+
+          const methodName = checkExists(getName(method.id));
+          // get() on InvitationsResource -> GetInvitationsResourceRequest
+          const requestTypeName = methodName.slice(0, 1).toUpperCase() + methodName.slice(1) + resourceInterfaceName + 'Request';
+
           const requestParameters: Record<string, gapi.client.discovery.JsonSchema> = { ...parameters, ...method.parameters };
-          const hasRequestRef = method.request && method.request['$ref'];
-          if (!(requestParameters.hasOwnProperty('resource') && hasRequestRef)) { // no resource param and no body at the same time -> generate x(request)
-            out.method(formatPropertyName(checkExists(getName(method.id))), [{
-              parameter: 'request',
-              type: (writer: TypescriptTextWriter) => {
-                writer.anonymousType(() => {
-                  forEachOrdered(requestParameters, (data, key) => {
-                    if (data.description) {
-                      writer.comment(formatComment(data.description));
-                    }
-                    writer.property(key, getType(data, schemas), data.required || false);
-                  });
-
-                  if (method.request && method.request['$ref']) {
-                    writer.comment('Request body');
-                    writer.property('resource', method.request['$ref'], true);
-                  }
-                });
-              },
-
-            }], getMethodReturn(method, schemas));
-          }
-          if (method.request && method.request['$ref']) { // has body -> generate x(request, body)
-            out.method(formatPropertyName(checkExists(getName(method.id))), [{
+          const requestBodyType = method.request && method.request['$ref'];
+
+          requestTypeWriters.push(() => {
+            out.interface(requestTypeName, (writer: TypescriptTextWriter) => {
+
+              forEachOrdered(requestParameters, (data, key) => {
+                if (data.description) {
+                  writer.comment(formatComment(data.description));
+                }
+                writer.property(key, getType(data, schemas), data.required || false);
+              });
+
+              // Only generate 'resource' property for body if the request parameters doesn't have an existing 'resource' property.
+              if (requestParameters.resource == null && requestBodyType) {
+                writer.comment('Request body');
+                writer.property('resource', requestBodyType, true);
+              }
+            });
+          });
+
+          const methodReturn = getMethodReturn(method, schemas);
+          if (requestParameters.resource != null && requestBodyType) {
+            // If the request has an existing 'resource' property, we drop the additional request body 'resource' parameter
```

Comment:
Added your duplicate `resource` property fix here. Do you remember which Google API had a non-body `resource` property in the request so I can test it out?

Reply:
Don't remember.. maybe, bigquery or something like that. Anyway, you could just generate all types and run DefinetlyTyped tests, all typings except for DialogFlow should pass. I'll review in more detail tomorrow, it's getting late here :)
Thanks for PR!